Receiver support binaryType ‘fragments’ …

guymguym opened the pull request:
- emit message fragments array to avoid copyfull concat

Note:
binaryType used to affect before only in WHATWG API emulation
and the optional conversion to ArrayBuffer was done inside MessageEvent.
However in this change the receiver emits messages according to binaryType
so now also EventEmitter listeners added with on(‘message’, data => {})
will get the data according to binaryType.

guymguym commented:
@lpinca Hi, would love to get your review.
See the Note I wrote above - let me know if you think this is a breaking change or not, and if you think it's a good direction.
Thanks

lpinca commented on lib/Constants.js:
```diff
@@ -3,3 +3,6 @@
 exports.EMPTY_BUFFER = Buffer.alloc(0);
 exports.NOOP = () => {};
+
+exports.DEFAULT_BINARY_TYPE = 'nodebuffer';
```
I would get rid of this and use `BINARY_TYPES[0]` instead where needed.

guymguym replied:
np

lpinca commented on lib/Receiver.js:
```diff
@@ -352,20 +354,28 @@ class Receiver {
 
       if (this.opcode === 2) {
-        this.onmessage(buf, { masked: this.masked, binary: true });
+        let data;
```
Please use `var` here due to possible deoptimizations.

guymguym replied:
np. 
For my education - would you mind explaining why this is a case for deoptimization?

lpinca replied:
I know for sure that [compund let assignment](https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax) are not optimizable on older V8 versions.

I also saw deoptimizations for non compound assignment when rewriting the parser so for now I prefer to stick with `var` when in doubt. This same is also done in Node.js core.

guymguym replied:
Thanks. That's good to know.

lpinca commented on lib/Receiver.js:
```diff
@@ -352,20 +354,28 @@ class Receiver {
+        if (this.binaryType === 'nodebuffer') {
+          data = bufferFromFragments(fragments, messageLength);
+        } else if (this.binaryType === 'arraybuffer') {
+          data = new Uint8Array(bufferFromFragments(fragments, messageLength)).buffer;
```
I'm not sure about the "semverness" of this change as previously conversion to `ArrayBuffer` was only done when using the `EventTarget` interface.

I like this as it makes sense and we also had a couple of issues opened because people were surprised to get a `Buffer` when they specified `binaryType` to be `arraybuffer` so this change can also be seen as a bug fix.

guymguym replied:
Yeah, it makes sense to me too, this is why I tried to suggest it this way.
If you want to play extra safe for the other folks then best release as a new major version.

lpinca commented on lib/Receiver.js:
```diff
@@ -352,20 +354,28 @@ class Receiver {
+          data = bufferFromFragments(fragments, messageLength);
+        } else if (this.binaryType === 'arraybuffer') {
+          data = new Uint8Array(bufferFromFragments(fragments, messageLength)).buffer;
+        } else { // if (this.binaryType === 'fragments') {
```
I don't think this comment is necessary. `data = fragments` is clear enough imo.

guymguym replied:
np

lpinca commented on lib/WebSocket.js:
```diff
@@ -98,10 +98,15 @@ class WebSocket extends EventEmitter {
-    } else {
-      throw new SyntaxError('unsupported binaryType: must be either "nodebuffer" or "arraybuffer"');
+    if (constants.BINARY_TYPES.indexOf(type) < 0) {
+      throw new SyntaxError('unsupported binaryType: must be either "' +
```
This is a breaking change as the error message is changed but I **think** there is not much code that relies on this. We can try to search for this error message on Github.

I think it's better to remove the error completely. The browser no longer throw this error if an invalid `binaryType` is specified. We should do the same and fail silently if an invalid `binaryType` is supplied.

guymguym replied:
In chrome it ignores unsupported values but with a console.warn (see below)
Do you think we should do console.warn too?

![image](https://cloud.githubusercontent.com/assets/1091422/23342806/666e9c2c-fc69-11e6-85d0-0d9196500b37.png)

lpinca replied:
Firefox doesn't warn, it just fails silently. What does WHATWG spec say?

lpinca replied:
https://html.spec.whatwg.org/multipage/comms.html#the-websocket-interface

> When a WebSocket object is created, its binaryType IDL attribute must be set to the string "blob". On getting, it must return the last value it was set to. On setting, the user agent must set the IDL attribute to the new value.

> Note: This attribute allows authors to control how binary data is exposed to scripts. By setting the attribute to "blob", binary data is returned in Blob form; by setting it to "arraybuffer", it is returned in ArrayBuffer form. User agents can use this as a hint for how to handle incoming binary data: if the attribute is set to "blob", it is safe to spool it to disk, and if it is set to "arraybuffer", it is likely more efficient to keep the data in memory. Naturally, user agents are encouraged to use more subtle heuristics to decide whether to keep incoming data in memory or not, e.g. based on how big the data is or how common it is for a script to change the attribute at the last minute. This latter aspect is important in particular because it is quite possible for the attribute to be changed after the user agent has received the data but before the user agent has fired the event for it.

This is all so it should be ok to just remove the error.

guymguym replied:
done

lpinca commented on test/Receiver.test.js:
```diff
@@ -827,4 +827,84 @@ describe('Receiver', function () {
+ * @param {Buffer[]} frags
+ * @private
+ */
+function addBinaryFragments (receiver, frags) {
```
Maybe we can get rid of this and use `Sender.frame()` if #1016 is merged. It will be easier to grok.

lpinca replied:
Merged #1016 so you can rebase and use that.

guymguym replied:
done

lpinca commented:
@guymguym I've requested some changes but LGTM.

guymguym commented:
@lpinca Fixed all comments. Thanks for the review!

lpinca commented:
Can you please run the parser benchmark and post the results before and after this change? Thanks.

guymguym commented:
Before:
```
$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.

$ node bench/parser.benchmark.js 
ping message x 1,516,912 ops/sec ±0.66% (90 runs sampled)
ping with no data x 14,433,226 ops/sec ±1.00% (88 runs sampled)
close message x 12,816,502 ops/sec ±1.34% (87 runs sampled)
masked text message (20 bytes) x 1,010,852 ops/sec ±1.13% (88 runs sampled)
binary data (125 bytes) x 1,284,809 ops/sec ±1.29% (90 runs sampled)
binary data (65535 bytes) x 186,970 ops/sec ±1.17% (87 runs sampled)
binary data (200 KiB) x 69,283 ops/sec ±1.10% (86 runs sampled)
binary data (1 MiB) x 12,406 ops/sec ±0.90% (89 runs sampled)

$ node bench/parser.benchmark.js 
ping message x 1,500,888 ops/sec ±0.68% (89 runs sampled)
ping with no data x 13,664,804 ops/sec ±1.78% (83 runs sampled)
close message x 12,325,815 ops/sec ±1.19% (85 runs sampled)
masked text message (20 bytes) x 945,751 ops/sec ±0.98% (91 runs sampled)
binary data (125 bytes) x 1,219,381 ops/sec ±1.06% (87 runs sampled)
binary data (65535 bytes) x 184,835 ops/sec ±1.39% (87 runs sampled)
binary data (200 KiB) x 69,053 ops/sec ±1.00% (86 runs sampled)
binary data (1 MiB) x 12,375 ops/sec ±0.96% (87 runs sampled)
```

After:
```
$ git checkout guymguym-receiver-fragments 
Switched to branch 'guymguym-receiver-fragments'
Your branch is up-to-date with 'origin/guymguym-receiver-fragments'.

$ node bench/parser.benchmark.js 
ping message x 1,537,659 ops/sec ±0.98% (89 runs sampled)
ping with no data x 14,070,824 ops/sec ±0.85% (87 runs sampled)
close message x 12,873,083 ops/sec ±1.19% (85 runs sampled)
masked text message (20 bytes) x 944,098 ops/sec ±0.89% (87 runs sampled)
binary data (125 bytes) x 1,404,478 ops/sec ±1.30% (85 runs sampled)
binary data (65535 bytes) x 188,281 ops/sec ±1.31% (84 runs sampled)
binary data (200 KiB) x 70,403 ops/sec ±1.05% (86 runs sampled)
binary data (1 MiB) x 12,387 ops/sec ±0.91% (87 runs sampled)

$ node bench/parser.benchmark.js 
ping message x 1,528,134 ops/sec ±0.96% (90 runs sampled)
ping with no data x 14,189,599 ops/sec ±0.87% (89 runs sampled)
close message x 12,886,198 ops/sec ±1.64% (88 runs sampled)
masked text message (20 bytes) x 1,048,379 ops/sec ±1.38% (89 runs sampled)
binary data (125 bytes) x 1,412,051 ops/sec ±0.91% (89 runs sampled)
binary data (65535 bytes) x 189,763 ops/sec ±1.09% (89 runs sampled)
binary data (200 KiB) x 70,405 ops/sec ±1.15% (86 runs sampled)
binary data (1 MiB) x 12,353 ops/sec ±0.86% (86 runs sampled)
```

guymguym commented:
Travis failed for node v4 but I think it's transient.
I don't have the option to re-trigger without commit, do you?

lpinca commented:
Yes, test seems to be flaky. The timeout is probably too short. I'll restart.

I'll keep this open for a while so others can weigh in. If there are no objections I'll merge this tomorrow.

lpinca commented:
Thanks!
